src/train.py

```diff
@@ -163,39 +261,86 @@ def optimizer_step(model, optimizer, args):
+    losses['tts_'].append(tts_loss.detach().cpu().item())
+    if use_discriminator:
+        losses['t_sp_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_sp_d'].append(s_d_loss.detach().cpu().item())
     return loss
 
-def train_ae_step(losses, model, batch, accum_steps):
+def train_ae_step(losses, model, batch, accum_steps, use_discriminator):
     batch = process_batch(batch)
-    t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
-    loss = (t_ae_loss + s_ae_loss) / accum_steps
+    if use_discriminator:
+        t_ae_loss, s_ae_loss, t_d_loss, s_d_loss = autoencoder_step(model, batch, use_discriminator)
+        loss = t_ae_loss + s_ae_loss + t_d_loss + s_d_loss
+    else:
+        t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
+        loss = t_ae_loss + s_ae_loss
+    loss = loss / accum_steps
     loss.backward()
 
     # Log losses
-    losses['t_ae'].append(t_ae_loss.detach().item())
-    losses['s_ae'].append(s_ae_loss.detach().item())
+    losses['t_ae'].append(t_ae_loss.detach().cpu().item())
+    losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+    if use_discriminator:
+        losses['t_ae_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_ae_d'].append(s_d_loss.detach().cpu().item())
     return loss
 
-def train_cm_step(losses, model, batch, accum_steps):
+def train_cm_step(losses, model, batch, accum_steps, use_discriminator):
+    # NOTE: do not use cross_model here bc need to take optimizer step inbetween
     batch = process_batch(batch)
-    t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
-    loss = (s_cm_loss + t_cm_loss) / accum_steps
+
+    if use_discriminator:
+        t_cm_loss, s_cm_loss, s_d_loss, cm_t_d_loss, t_d_loss, cm_s_d_loss = crossmodel_step(model, batch, use_discriminator)
+        loss = s_cm_loss + t_cm_loss + s_d_loss + cm_t_d_loss + t_d_loss + cm_s_d_loss
+    else:
+        t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
+        loss = s_cm_loss + t_cm_loss
+    loss = loss / accum_steps
     loss.backward()
 
     # Log losses
-    losses['s_cm'].append(s_cm_loss.detach().item())
-    losses['t_cm'].append(t_cm_loss.detach().item())
+    losses['s_cm'].append(s_cm_loss.detach().cpu().item())
+    losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['s_cm_d'].append(s_d_loss.detach().cpu().item())
+        losses['t_cm_d'].append(t_d_loss.detach().cpu().item())
+        losses['cm_s_cm_d'].append(cm_s_d_loss.detach().cpu().item())
+        losses['cm_t_cm_d'].append(cm_t_d_loss.detach().cpu().item())
     return loss
 
+def train_discriminator_step(losses, model, batch, accum_steps):
+    t_d_loss, s_d_loss = discriminator_step(model, batch)
+
+    loss = (t_d_loss + s_d_loss) / accum_steps
+    loss.backward()
+
+    losses['t_d'].append(t_d_loss.detach().cpu().item())
+    losses['s_d'].append(s_d_loss.detach().cpu().item())
+    return loss
+
+def freeze_model_parameters(model):
```

Comment:
Can we use .eval() and .train() to achieve this freeze/unfreeze

Reply:
No, unfortunately torch doesn't like calling eval() on the encoders and decoders and then trying to do an optimizer step.

Reply:
Oh, I see - this is a result of all parameters using the same optimizer. Well that is unfortunate 😞